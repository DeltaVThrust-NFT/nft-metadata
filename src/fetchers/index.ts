import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
