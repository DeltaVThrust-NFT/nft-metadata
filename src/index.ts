export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
