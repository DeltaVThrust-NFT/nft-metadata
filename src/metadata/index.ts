export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
