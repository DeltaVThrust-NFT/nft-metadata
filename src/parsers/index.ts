import { getAddress } from '@ethersproject/address'
import { parseGenericMetadata } from './openseaMetadataParser'
import {
  ART_BLOCKS_CURATED_TOKEN_ADDRESS,
  ART_BLOCKS_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { parseZoraMetadata } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { parseArtblocksMetadata } from './artblocksMetadataParser'

export function parserLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    case ART_BLOCKS_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case ART_BLOCKS_CURATED_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return parseZoraMetadata
    default:
      return parseGenericMetadata
  }
}
