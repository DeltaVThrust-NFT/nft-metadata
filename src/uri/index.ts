export { getAlternateContractCall } from './contract'
export {
  fetchMimeType,
  fetchURI,
  fetchWithRetries,
  fetchWithTimeout,
  parseDataUri,
} from './fetch'
export { isIPFS, getIPFSUrl, getPrivateGateway } from './ipfs'
export { getStaticURI, getURIData } from './static'
