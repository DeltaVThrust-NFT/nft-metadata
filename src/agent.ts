import { JsonRpcProvider } from '@ethersproject/providers'
import { fetchOwnerOf, getSupportedInterfaces } from './utils/contracts'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { Fetcher, fetcherLookup, FetcherResponse } from './fetchers'
import { fetchMetadata, fetchMimeType } from './utils/fetch'
import { getIPFSUrl } from './utils/ipfs'
import { Parser, parserLookup } from './parsers'

export interface TokenPKDto {
  networkId: number
  tokenAddress: string
  tokenId: string
}

export interface NftMetadata extends TokenPKDto {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string
  creatorAddress?: string

  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string

  contentURL?: string
  contentURLMimeType?: string

  imageURL?: string
  imageURLMimeType?: string

  externalLink?: string

  attributes?: Record<string, any>[]
}

export enum Network {
  MAINNET = 1,
  RINKEBY = 4,
}

export type AgentNetworkConfiguration = {
  [key: number]: JsonRpcProvider
}

export type AgentConfiguration = {
  ipfsGateway?: string
  fetchTimeout?: number
  providers: AgentNetworkConfiguration
  parsers?: {
    [key: string]: Parser
  }
  fetchers?: {
    [key: string]: Fetcher
  }
}

export type TokenContractData = FetcherResponse & TokenPKDto & Record<any, any>
export type TokenData = { metadata: Record<any, any>; contentType: string }

export class Agent {
  constructor(private readonly options: AgentConfiguration) {}

  private getProviderForNetwork(number: Network): JsonRpcProvider {
    const provider = this.options.providers?.[number]
    if (!provider) {
      throw new Error('No provider for network')
    }
    return provider
  }

  public async supportedInterfaces(networkId: Network, tokenAddress: string) {
    const provider = this.getProviderForNetwork(networkId)
    return getSupportedInterfaces(provider, tokenAddress)
  }

  public async fetchContractData(
    networkId: Network,
    tokenAddress: string,
    tokenId: string,
  ) {
    const provider = this.getProviderForNetwork(networkId)
    const { erc721, erc721Metadata, erc721Enumerable } =
      await this.supportedInterfaces(networkId, tokenAddress)

    if (!erc721) {
      throw new Error(
        `${tokenAddress} on network: ${networkId} does not support ERC721 standard`,
      )
    }

    const erc721Contract = Erc721Factory.connect(tokenAddress, provider)
    const ownerAddress = await fetchOwnerOf(
      erc721Contract,
      tokenId,
      // Can catch if contract can make a total supply call
      // to handle open zeppelin burn
      erc721Enumerable,
    )

    let tokenURI: string | undefined
    if (erc721Metadata) {
      tokenURI = await erc721Contract.tokenURI(tokenId)
    }

    const baseData = {
      networkId,
      tokenAddress,
      tokenId,
      tokenURI,
      ownerAddress,
    }

    // bespoke fetches for other underlying data
    const fetcher = fetcherLookup(tokenAddress, this.options.fetchers)
    if (fetcher) {
      const data = await fetcher({
        provider,
        tokenAddress,
        tokenId,
      })

      return {
        ...baseData,
        ...data,
      }
    }

    return baseData
  }

  public async parseMetadata(
    contractData: TokenContractData,
    tokenData: TokenData,
  ) {
    const metadata = tokenData.metadata

    let meta: Partial<NftMetadata> = {}

    if (contractData.tokenURI) {
      meta.tokenURL = getIPFSUrl(
        contractData.tokenURI,
        this.options.ipfsGateway,
      )
      meta.tokenURLMimeType = tokenData.contentType
    }

    if (metadata.image) {
      meta.imageURL = getIPFSUrl(metadata.image, this.options.ipfsGateway)
      meta.contentURL = meta.imageURL
    }

    if (metadata.animation_url) {
      meta.contentURL = getIPFSUrl(
        metadata.animation_url,
        this.options.ipfsGateway,
      )
    }

    const {
      name,
      description,
      attributes,
      traits,
      external_url: externalURL,
      mimeType,
    } = metadata

    if (meta.contentURL) {
      meta.contentURLMimeType = await fetchMimeType(
        meta.contentURL,
        {
          timeout: this.options.fetchTimeout,
        },
        mimeType,
      )
    }

    if (meta.imageURL && meta.imageURL !== meta.contentURL) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.options.fetchTimeout,
      })
    } else {
      delete meta.imageURL
    }

    const baseMeta = {
      metadata,
      ...meta,
      ...(name && { name }),
      ...(description && { description }),
      ...(attributes && { attributes }),
      ...(traits && { attributes: traits }),
      ...(externalURL && { externalURL }),
    }

    // Return any extra fields unparseable by defaults
    // @TODO - @ethandaya move this behaviour to a dynamic registry
    const parser = parserLookup(contractData.tokenAddress, this.options.parsers)
    if (parser) {
      const meta = await parser({
        provider: this.getProviderForNetwork(contractData.networkId),
        contractData,
        tokenData,
        ipfsGateway: this.options.ipfsGateway,
        fetchTimeout: this.options.fetchTimeout,
        baseMeta,
      })
      return {
        ...baseMeta,
        ...meta,
      }
    }

    return baseMeta
  }

  public async fetchAndParseTokenData(
    networkId: number,
    tokenAddress: string,
    tokenId: string,
  ) {
    const contractData = await this.fetchContractData(
      networkId,
      tokenAddress,
      tokenId,
    )

    if (!contractData.tokenURI) {
      return { ...contractData }
    }

    const tokenData = await fetchMetadata(contractData.tokenURI, {
      timeout: this.options.fetchTimeout,
      ipfsGateway: this.options.ipfsGateway,
    })

    const metadata = await this.parseMetadata(contractData, tokenData)

    return {
      ...contractData,
      ...metadata,
    }
  }
}
