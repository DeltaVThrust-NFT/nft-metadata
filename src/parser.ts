import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
