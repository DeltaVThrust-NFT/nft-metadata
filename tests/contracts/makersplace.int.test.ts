import { Agent, MAKERSPLACE_TOKEN_ADDRESS, Network } from '../../src'
import { testProvider } from '../setupProvider'
import { isAddress } from '@ethersproject/address'

const IMAGE_CRITERIA = {
  input: {
    tokenId: '100',
    tokenAddress: MAKERSPLACE_TOKEN_ADDRESS,
    networkId: 1,
  },
}
const VIDEO_CRITERIA = {
  input: {
    tokenId: '63253',
    tokenAddress: MAKERSPLACE_TOKEN_ADDRESS,
    networkId: 1,
  },
}

describe('Makersplace ERC721', () => {
  const parser = new Agent({
    providers: {
      [Network.MAINNET]: testProvider,
    },
    ipfsGateway: 'https://ipfsgateway.makersplace.com',
    fetchTimeout: 60000,
  })

  beforeEach(() => {
    jest.setTimeout(120000)
  })

  it(`should be able to fetch and parse metadata for makersplace image — token id: ${IMAGE_CRITERIA.input.tokenId} on network: ${IMAGE_CRITERIA.input.networkId}`, async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenData(
      1,
      IMAGE_CRITERIA.input.tokenAddress,
      IMAGE_CRITERIA.input.tokenId,
    )
    expect(meta).toMatchSnapshot()
    expect(isAddress(ownerAddress)).toBeTruthy()
  })

  it(`should be able to fetch and parse metadata for makersplace video — token id: ${VIDEO_CRITERIA.input.tokenId} on network: ${IMAGE_CRITERIA.input.networkId}`, async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenData(
      1,
      VIDEO_CRITERIA.input.tokenAddress,
      VIDEO_CRITERIA.input.tokenId,
    )
    expect(meta).toMatchSnapshot()
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
