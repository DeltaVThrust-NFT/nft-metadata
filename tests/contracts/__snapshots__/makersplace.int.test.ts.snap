// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Makersplace ERC721 should be able to fetch and parse metadata for makersplace image — token id: 100 on network: 1 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "Creator",
      "value": "Paul Crudgington",
    },
  ],
  "contentURL": "https://ipfsgateway.makersplace.com/ipfs/QmPPxhZmDX4TiXoj6EykoU8k23YUq3XDgYrrwT55hB3kt8",
  "contentURLMimeType": "image/jpeg",
  "description": "This is a really popular spot at The Tate Modern (London) for photographers. The brutalist concrete architecture and sweeping lines make it a dream to shoot, and this is my take on that scene.
The angle at which the image is made and the processing in black and white, really helps to accentuate the curves and lines and the presence of the lone person gives the sense of scale.",
  "metadata": Object {
    "attributes": Array [
      Object {
        "trait_type": "Creator",
        "value": "Paul Crudgington",
      },
    ],
    "description": "This is a really popular spot at The Tate Modern (London) for photographers. The brutalist concrete architecture and sweeping lines make it a dream to shoot, and this is my take on that scene.
The angle at which the image is made and the processing in black and white, really helps to accentuate the curves and lines and the presence of the lone person gives the sense of scale.",
    "imageUrl": "https://ipfsgateway.makersplace.com/ipfs/QmPPxhZmDX4TiXoj6EykoU8k23YUq3XDgYrrwT55hB3kt8",
    "name": "Tate Modern",
    "properties": Object {
      "created_at": Object {
        "description": "2019-03-15T10:53:01.353471+00:00",
        "type": "datetime",
      },
      "description": Object {
        "description": "This is a really popular spot at The Tate Modern (London) for photographers. The brutalist concrete architecture and sweeping lines make it a dream to shoot, and this is my take on that scene.
The angle at which the image is made and the processing in black and white, really helps to accentuate the curves and lines and the presence of the lone person gives the sense of scale.",
        "type": "string",
      },
      "digital_media_signature": Object {
        "description": "d6bf9a289b1f8845190220a16be2fe0ea31f7c6a7ead6f59af6f459a093e1f9b",
        "type": "string",
      },
      "digital_media_signature_type": Object {
        "description": "SHA-256",
        "type": "string",
      },
      "name": Object {
        "description": "Tate Modern",
        "type": "string",
      },
      "preview_media_file": Object {
        "description": "https://ipfsgateway.makersplace.com/ipfs/QmPPxhZmDX4TiXoj6EykoU8k23YUq3XDgYrrwT55hB3kt8",
        "type": "string",
      },
      "preview_media_file_type": Object {
        "description": "jpg",
        "type": "string",
      },
      "total_supply": Object {
        "description": 10,
        "type": "int",
      },
    },
    "title": "Tate Modern",
    "type": "object",
  },
  "name": "Tate Modern",
  "networkId": 1,
  "tokenAddress": "0x2A46f2fFD99e19a89476E2f62270e0a35bBf0756",
  "tokenId": "100",
  "tokenURI": "ipfs://ipfs/QmQZ66nmWPRFTqfuxVuAMHKb57xwhDZfGkSDGd7MJjxbWm",
  "tokenURL": "https://ipfsgateway.makersplace.com/ipfs/QmQZ66nmWPRFTqfuxVuAMHKb57xwhDZfGkSDGd7MJjxbWm",
  "tokenURLMimeType": "application/json",
}
`;

exports[`Makersplace ERC721 should be able to fetch and parse metadata for makersplace video — token id: 63253 on network: 1 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "Creator",
      "value": "Reema",
    },
  ],
  "contentURL": "https://ipfsgateway.makersplace.com/ipfs/QmbhDKcf4fqtjXQvWmSyJT2pU1pYwCoa1wWVkTurBvkBcV",
  "contentURLMimeType": "video/mp4",
  "description": "We are the cosmos made conscious. We are consciousness incarnated in stardust. \\"We inhabit a universe where atoms are made in the centers of stars; where each second a thousand suns are born; where the raw material for biological evolution is sometimes made by the explosion of a star halfway across the Milky Way; where a thing as beautiful as a galaxy is formed a hundred billion times . We are the local embodiment of a Cosmos grown to self-awareness. We have begun to contemplate our origins: stardust pondering the stars; organized assemblages of ten billion billion billion atoms considering the evolution of atoms; tracing the long journey by which, here at least, consciousness arose. Not only are we made of stardust; we’re made of elements so fundamental to the universe they were created in the first moments of the Big Bang. The history inside of us stretches back further than we knew and our future remains inescapably among the stars. We are the universe trying to understand itself\\"",
  "imageURL": "https://ipfsgateway.makersplace.com/ipfs/QmeSH3ap2VCN6KgL4U8ebvrYrWBKPYffJsAbAUiHk9qSik",
  "imageURLMimeType": "image/jpeg",
  "metadata": Object {
    "attributes": Array [
      Object {
        "trait_type": "Creator",
        "value": "Reema",
      },
    ],
    "description": "We are the cosmos made conscious. We are consciousness incarnated in stardust. \\"We inhabit a universe where atoms are made in the centers of stars; where each second a thousand suns are born; where the raw material for biological evolution is sometimes made by the explosion of a star halfway across the Milky Way; where a thing as beautiful as a galaxy is formed a hundred billion times . We are the local embodiment of a Cosmos grown to self-awareness. We have begun to contemplate our origins: stardust pondering the stars; organized assemblages of ten billion billion billion atoms considering the evolution of atoms; tracing the long journey by which, here at least, consciousness arose. Not only are we made of stardust; we’re made of elements so fundamental to the universe they were created in the first moments of the Big Bang. The history inside of us stretches back further than we knew and our future remains inescapably among the stars. We are the universe trying to understand itself\\"",
    "imageUrl": "https://ipfsgateway.makersplace.com/ipfs/QmeSH3ap2VCN6KgL4U8ebvrYrWBKPYffJsAbAUiHk9qSik",
    "name": "We Are Made of Stardust",
    "properties": Object {
      "created_at": Object {
        "description": "2021-06-16T18:31:20.779279+00:00",
        "type": "datetime",
      },
      "description": Object {
        "description": "We are the cosmos made conscious. We are consciousness incarnated in stardust. \\"We inhabit a universe where atoms are made in the centers of stars; where each second a thousand suns are born; where the raw material for biological evolution is sometimes made by the explosion of a star halfway across the Milky Way; where a thing as beautiful as a galaxy is formed a hundred billion times . We are the local embodiment of a Cosmos grown to self-awareness. We have begun to contemplate our origins: stardust pondering the stars; organized assemblages of ten billion billion billion atoms considering the evolution of atoms; tracing the long journey by which, here at least, consciousness arose. Not only are we made of stardust; we’re made of elements so fundamental to the universe they were created in the first moments of the Big Bang. The history inside of us stretches back further than we knew and our future remains inescapably among the stars. We are the universe trying to understand itself\\"",
        "type": "string",
      },
      "digital_media_signature": Object {
        "description": "da6a09de80950dfbb2ae9bbd00db117e08d9ee2d93a6d4445b9490a27539310e",
        "type": "string",
      },
      "digital_media_signature_type": Object {
        "description": "SHA-256",
        "type": "string",
      },
      "name": Object {
        "description": "We Are Made of Stardust",
        "type": "string",
      },
      "preview_media_file": Object {
        "description": "https://ipfsgateway.makersplace.com/ipfs/QmeSH3ap2VCN6KgL4U8ebvrYrWBKPYffJsAbAUiHk9qSik",
        "type": "string",
      },
      "preview_media_file2": Object {
        "description": "https://ipfsgateway.makersplace.com/ipfs/QmbhDKcf4fqtjXQvWmSyJT2pU1pYwCoa1wWVkTurBvkBcV",
        "type": "string",
      },
      "preview_media_file2_type": Object {
        "description": "mp4",
        "type": "string",
      },
      "preview_media_file_type": Object {
        "description": "jpg",
        "type": "string",
      },
      "total_supply": Object {
        "description": 1,
        "type": "int",
      },
    },
    "title": "We Are Made of Stardust",
    "type": "object",
  },
  "name": "We Are Made of Stardust",
  "networkId": 1,
  "tokenAddress": "0x2A46f2fFD99e19a89476E2f62270e0a35bBf0756",
  "tokenId": "63253",
  "tokenURI": "ipfs://ipfs/QmczR8BYHGsfMjiQ988mhrQo4Femozf4RoVsqu8PjxFMNU",
  "tokenURL": "https://ipfsgateway.makersplace.com/ipfs/QmczR8BYHGsfMjiQ988mhrQo4Femozf4RoVsqu8PjxFMNU",
  "tokenURLMimeType": "application/json",
}
`;
